components/crypto/crypto-ecdh-impl/src/main/kotlin/net/corda/crypto/ecdh/impl/SharedSecretOps.kt

```diff
@@ -0,0 +1,131 @@
+package net.corda.crypto.ecdh.impl
+
+import net.corda.crypto.core.concatByteArrays
+import net.corda.crypto.core.toByteArray
+import net.corda.crypto.ecdh.AES_KEY_ALGORITHM
+import net.corda.crypto.ecdh.AES_KEY_SIZE_BYTES
+import net.corda.crypto.ecdh.AES_PROVIDER
+import net.corda.crypto.ecdh.GCM_NONCE_LENGTH
+import net.corda.crypto.ecdh.GCM_TAG_LENGTH
+import net.corda.crypto.ecdh.GCM_TRANSFORMATION
+import org.bouncycastle.crypto.generators.HKDFBytesGenerator
+import org.bouncycastle.crypto.params.HKDFParameters
+import org.bouncycastle.jcajce.provider.util.DigestFactory
+import java.security.PublicKey
+import java.util.concurrent.ConcurrentLinkedQueue
+import java.util.concurrent.atomic.AtomicInteger
+import javax.crypto.Cipher
+import javax.crypto.spec.GCMParameterSpec
+import javax.crypto.spec.SecretKeySpec
+import kotlin.experimental.xor
+
+object SharedSecretOps {
+    private val nonceCounter = AtomicInteger(0)
+
+    private val pool = ConcurrentLinkedQueue<Cipher>()
+
+    private fun withGcmCipherInstance(block: Cipher.() -> ByteArray): ByteArray {
+        val cipher = pool.poll()
+            ?: Cipher.getInstance(GCM_TRANSFORMATION, AES_PROVIDER)
+        try {
+            return cipher.block()
+        } finally {
+            pool.offer(cipher)
+        }
+    }
+
+    fun encrypt(
+        digestName: String,
+        salt: ByteArray,
+        info: ByteArray,
+        publicKey: PublicKey,
+        otherPublicKey: PublicKey,
+        plainText: ByteArray,
+        aad: ByteArray?,
+        deriveSharedSecret: (PublicKey) -> ByteArray
+    ): ByteArray {
+        val currentNonceCounter = nonceCounter.getAndIncrement().toByteArray()
+        val data = derive(
+            digestName = digestName,
+            salt = salt,
+            info = info,
+            publicKey = publicKey,
+            currentNonceCounter = currentNonceCounter,
+            deriveSharedSecret = deriveSharedSecret
+        )
+        val finalAad = concatByteArrays(
+            aad ?: ByteArray(0), publicKey.encoded, otherPublicKey.encoded
+        )
+        return withGcmCipherInstance {
+            init(Cipher.ENCRYPT_MODE, data.key, GCMParameterSpec(GCM_TAG_LENGTH * 8, data.nonce))
+            updateAAD(finalAad)
+            concatByteArrays(currentNonceCounter, doFinal(plainText))
+        }
+    }
+
+    fun decrypt(
+        digestName: String,
+        salt: ByteArray,
+        info: ByteArray,
+        publicKey: PublicKey,
+        otherPublicKey: PublicKey,
+        cipherText: ByteArray,
+        aad: ByteArray?,
+        deriveSharedSecret: (PublicKey) -> ByteArray
+    ): ByteArray {
+        val currentNonceCounter = cipherText.sliceArray(0 until GCM_NONCE_LENGTH)
```

Comment:
```suggestion
        val currentNonceCounter = cipherText.sliceArray(0 until 4)
```
or have a separate constant for the nonce counter size?

I think the way it is now it works accidentally, because XOR is performed only on the first 4 bytes..

Reply:
That is changed to use Long.SIZE_BYTES, not commited yet